modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2087,6 +2088,35 @@ def concat(self, axis, others, how, sort):
             new_lengths = None
             new_widths = None
         new_partitions = self._partition_mgr_cls.concat(axis, left_parts, right_parts)
+        shape = new_partitions.shape
+        # In the case when the number of partitions significantly differs from the value
+        # specified by `NPartitions`, we can get a significant slowdown in subsequent operations.
+        desired_partition_count = NPartitions.get() * 2
+        if shape[axis] >= desired_partition_count:
+            from modin.core.storage_formats.pandas.utils import (
+                compute_default_axes_lengths,
+            )
+
+            new_partitions = self._partition_mgr_cls.map_axis_partitions(
```

Comment:
If it's more than `NPartitions.get() * 2` can we just combine a subset of the partitions instead of doing a full `map_axis_partitions`? We don't need the partitions to be perfectly aligned. What do you think?

Reply:
This can speed up the execution of one repartitioning operation, but the number of repartitions will grow and, as a result, the gain in execution speed may not be significant, on the other hand, the implementation complexity will increase.

It also seems to me that making a partial repartition deferred is also more difficult.

Reply:
This "partial combination" was what I proposed in https://github.com/modin-project/modin/pull/3720#issuecomment-975571059, but I totally agree with @anmyachev here that implementing that is much harder than just re-combining the whole axis (as it's a totally new operation I think).

Reply:
It's possible to do with current mechanisms (creating "column partitions" that return 1 split):

```python
step = 2  # some multiple of how many to combine
new_partitions = [self._partition_mgr_cls.map_axis_partitions(
    axis=axis,
    partitions=new_partitions[i:i+step],
    map_func=lambda df: df,
    keep_partitioning=False,
    lengths=[len(self.axes[axis])],
    enumerate_partitoins=False,
) for i in range(0, len(partitions), step)]
```

Does this make sense?

Reply:
In this case I propose to compute the `step` by dividing real size by desired number of partitions and rounding it down.

We can try to see how it goes in terms of speed.

Reply:
will try

Reply:
It is ugly, but this code block could potentially lead to a better abstraction for "merged partitions"

Reply:
Note that in your example the partitions were already pretty unbalanced 😄 not that balancing is a bad idea, but I won't be putting too much into it - who knows what other masks or anything a user would do that would unbalance the thing further?..

Reply:
However, this implementation will not speed up the custom workload.

Reply:
Thanks @anmyachev I agree, `step=2` could be a better heuristic. We can also "split" the larger partitions logically by using `mask` to make things a bit more balanced without bloating memory.

What custom workload are you optimizing for? Is it somewhere in this thread?

Reply:
The workload itself cannot be published here, but it consists with many `concat` operations, after which there is `to_csv` call. There is an approximate benchmark at the PR top.

Reply:
@anmyachev Do you have thoughts on using the `mask` logic combined with the logic above to avoid `full_axis` logic rebalancing here?

Reply:
@devin-petersohn let me try

Reply:
@devin-petersohn I implement the heuristic, that I talking before - "Estimating the number of columns in partitions, and changing the step based on". It should be better than use `mask` here.